readthedocs/core/views/serve.py

```diff
@@ -269,3 +272,105 @@ def robots_txt(request, project):
+    in the sitemap.
+
+    :param request: Django request object
+    :param project: Project instance to generate the sitemap
+
+    :returns: response with the ``sitemap.xml`` template rendered
+
+    :rtype: django.http.HttpResponse
+    """
+    def priorities_generator():
+        """
+        Generator returning ``priority`` needed by sitemap.xml.
+
+        It generates values from 1 to 0.1 by decreasing in 0.1 on each
+        iteration. After 0.1 is reached, it will keep returning 0.1.
+        """
+        priorities = [1, 0.9, 0.8, 0.7, 0.6, 0.5, 0.4, 0.3, 0.2]
+        yield from itertools.chain(priorities, itertools.repeat(0.1))
+
+    def changefreqs_generator():
+        """
+        Generator returning ``changefreq`` needed by sitemap.xml.
+
+        It returns ``daily`` on first iteration, then ``weekly`` and then it
+        will return always ``monthly``.
+
+        We are using ``monthly`` as last value because ``never`` is too
+        aggressive. If the tag is removed and a branch is created with the same
+        name, we will want bots to revisit this.
+        """
+        changefreqs = ['daily', 'weekly']
+        yield from itertools.chain(changefreqs, itertools.repeat('monthly'))
+
+    if project.privacy_level == constants.PRIVATE:
+        raise Http404
+
+    sorted_versions = sort_version_aware(
+        project.versions.filter(
+            active=True,
+            privacy_level=constants.PUBLIC,
+        ),
+    )
+
+    versions = []
+    for version, priority, changefreq in zip(
+            sorted_versions, priorities_generator(), changefreqs_generator()):
+        element = {
+            'loc': version.get_subdomain_url(),
+            'priority': priority,
+            'changefreq': changefreq,
+            'languages': [],
+        }
+
+        # Version can be enabled, but not ``built`` yet. We want to show the
+        # link without a ``lastmod`` attribute
+        last_build = version.builds.order_by('-date').first()
+        if last_build:
+            element['lastmod'] = last_build.date.isoformat()
+
+        if project.translations.exists():
+            for translation in project.translations.all():
+                href = project.get_docs_url(
+                    version_slug=version.slug,
+                    lang_slug=translation.language,
+                    private=version.privacy_level == constants.PRIVATE,
+                )
+                element['languages'].append({
+                    'hreflang': translation.language,
+                    'href': href,
+                })
+
+            # Add itself also as protocol requires
+            element['languages'].append({
+                'hreflang': project.language,
+                'href': element['loc'],
+            })
+
+        versions.append(element)
+
+    context = {
+        'versions': versions,
+    }
+    return render(request, 'sitemap.xml', context, content_type='application/xml')
```

Comment:
Any reason you went with a template instead of Django's builtin sitemap framework? https://docs.djangoproject.com/en/1.11/ref/contrib/sitemaps/

The framework might have some advantages in case we ever get so large we need to break up sitemaps into multiple files.

Reply:
Ignorance.

That said and after reading the documentation, I'm not sure how to:

* get the project from inside the Sitemap class (I think it doesn't have access to the `request` object --where we could check for `.slug` property, for example)
* generate different locations for translations that are not based on `LANGUAGE` variable but on `project.translations`

`Sitemap` objects look pretty clean and clear but I'd need some help on those two things to be able to make it work using the framework.